File: trainer/src/semantscript_trainer/semantic_json.py
"""Compiler-compatible ``semantscript-semantic-json`` version 1 encoding."""

from __future__ import annotations

import hashlib
import json
import math
import struct

type JsonValue = bool | int | float | str | list[JsonValue] | dict[str, JsonValue] | None
type TypedJsonNode = list[str | bool | int | list[TypedJsonNode] | list[list[str | TypedJsonNode]]]

_DOMAIN = "semantscript-semantic-json"
_VERSION = 1


def semantic_json_bytes(value: JsonValue, /) -> bytes:
    """Encode a JSON value using the compiler's typed semantic JSON v1 format."""

    return semantic_json_string(value).encode("utf-8")


def semantic_json_string(value: JsonValue, /) -> str:
    """Return the exact diagnostic string emitted by the TypeScript compiler."""

    document = [_DOMAIN, _VERSION, _to_typed_node(value)]
    return json.dumps(document, ensure_ascii=False, separators=(",", ":"))


def semantic_json_sha256(value: JsonValue, /) -> str:
    """Return the lowercase SHA-256 digest of typed semantic JSON v1 bytes."""

    return hashlib.sha256(semantic_json_bytes(value)).hexdigest()


def _to_typed_node(value: JsonValue) -> TypedJsonNode:
    if value is None:
        return ["null"]
    if type(value) is bool:
        return ["boolean", value]
    if type(value) in (int, float):
        return ["number", _number_to_binary64_hex(value)]
    if type(value) is str:
        _assert_unicode_scalar_string(value)
        return ["string", value]
    if type(value) is list:
        return ["array", [_to_typed_node(item) for item in value]]
    if type(value) is dict:
        entries: list[list[str | TypedJsonNode]] = []
        for name, entry_value in value.items():
            if type(name) is not str:
                raise TypeError("semantic JSON object keys must be strings")
            _assert_unicode_scalar_string(name)
            entries.append([name, _to_typed_node(entry_value)])
        entries.sort(key=lambda entry: entry[0].encode("utf-8"))
        return ["object", entries]
    raise TypeError("semantic JSON values must contain only JSON-compatible data")


def _number_to_binary64_hex(value: int | float) -> str:
    try:
        binary64 = float(value)
    except OverflowError as error:
        raise TypeError("semantic JSON numbers must be finite") from error
    if not math.isfinite(binary64):
        raise TypeError("semantic JSON numbers must be finite")
    return struct.pack(">d", binary64).hex()


def _assert_unicode_scalar_string(value: str) -> None:
    if any(0xD800 <= ord(character) <= 0xDFFF for character in value):
        raise TypeError("semantic JSON strings cannot contain unpaired UTF-16 surrogates")


__all__ = [
    "JsonValue",
    "semantic_json_bytes",
    "semantic_json_sha256",
    "semantic_json_string",
]

File: compiler/src/semantic-json.ts
const encoder = new TextEncoder();

export type JsonValue =
  | null
  | boolean
  | number
  | string
  | readonly JsonValue[]
  | { readonly [name: string]: JsonValue };

type TypedJsonNode =
  | readonly ["null"]
  | readonly ["boolean", boolean]
  | readonly ["number", string]
  | readonly ["string", string]
  | readonly ["array", readonly TypedJsonNode[]]
  | readonly ["object", readonly (readonly [string, TypedJsonNode])[]];

export function semanticJsonBytes(value: unknown): Uint8Array {
  return encoder.encode(semanticJsonString(value));
}

export function semanticJsonString(value: unknown): string {
  return JSON.stringify(["semantscript-semantic-json", 1, toTypedNode(value)]);
}

export function compareBytes(left: Uint8Array, right: Uint8Array): number {
  const length = Math.min(left.length, right.length);

  for (let index = 0; index < length; index += 1) {
    const difference = (left[index] ?? 0) - (right[index] ?? 0);

    if (difference !== 0) {
      return difference;
    }
  }

  return left.length - right.length;
}

export function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes, (byte) => byte.toString(16).padStart(2, "0")).join(
    "",
  );
}

function toTypedNode(value: unknown): TypedJsonNode {
  if (value === null) {
    return ["null"];
  }

  if (typeof value === "boolean") {
    return ["boolean", value];
  }

  if (typeof value === "number") {
    if (!Number.isFinite(value)) {
      throw new TypeError("semantic JSON numbers must be finite");
    }

    return ["number", numberToBinary64Hex(value)];
  }

  if (typeof value === "string") {
    assertUnicodeScalarString(value);
    return ["string", value];
  }

  if (Array.isArray(value)) {
    return ["array", value.map(toTypedNode)];
  }

  if (typeof value !== "object") {
    throw new TypeError(
      "semantic JSON values must contain only JSON-compatible data",
    );
  }

  const entries = Object.entries(value)
    .map(([name, entryValue]) => {
      assertUnicodeScalarString(name);
      return [name, toTypedNode(entryValue)] as const;
    })
    .sort(([left], [right]) =>
      compareBytes(encoder.encode(left), encoder.encode(right)),
    );
  return ["object", entries];
}

function numberToBinary64Hex(value: number): string {
  const buffer = new ArrayBuffer(8);
  new DataView(buffer).setFloat64(0, value, false);
  return bytesToHex(new Uint8Array(buffer));
}

function assertUnicodeScalarString(value: string): void {
  for (let index = 0; index < value.length; index += 1) {
    const code = value.charCodeAt(index);

    if (code >= 0xd800 && code <= 0xdbff) {
      const next = value.charCodeAt(index + 1);

      if (!(next >= 0xdc00 && next <= 0xdfff)) {
        throw new TypeError(
          "semantic JSON strings cannot contain unpaired UTF-16 surrogates",
        );
      }

      index += 1;
    } else if (code >= 0xdc00 && code <= 0xdfff) {
      throw new TypeError(
        "semantic JSON strings cannot contain unpaired UTF-16 surrogates",
      );
    }
  }
}
